build/cmake/ResSearch.cmake

```diff
@@ -0,0 +1,57 @@
+include(CheckSymbolExists)
+include(CMakePushCheckState)
+
+cmake_push_check_state(RESET)
+
+# The name of the library that performs name resolution, suitable for giving to the "-l" link flag
+set(RESOLVE_LIB_NAME)
+# If TRUE, then the C runtime provides the name resolution that we need
+set(resolve_is_libc FALSE)
+
+if(WIN32)
+    set(RESOLVE_LIB_NAME Dnsapi)
+    set(_MONGOC_HAVE_DNSAPI 1)
+else()
+    # Try to find the search functions for various configurations.
+    # Headers required by minimum on the strictest system: (Tested on FreeBSD 13)
+    set(resolve_headers netinet/in.h sys/types.h arpa/nameser.h resolv.h)
+    set(CMAKE_REQUIRED_LIBRARIES resolv)
+    check_symbol_exists(res_nsearch "${resolve_headers}" _MONGOC_HAVE_RES_NSEARCH_RESOLV)
+    check_symbol_exists(res_search "${resolve_headers}" _MONGOC_HAVE_RES_SEARCH_RESOLV)
+    check_symbol_exists(res_ndestroy "${resolve_headers}" _MONGOC_HAVE_RES_NDESTROY_RESOLV)
+    check_symbol_exists(res_nclose "${resolve_headers}" _MONGOC_HAVE_RES_NCLOSE_RESOLV)
+    if((_MONGOC_HAVE_RES_NSEARCH_RESOLV OR _MONGOC_HAVE_RES_SEARCH_RESOLV)
+        AND (_MONGOC_HAVE_RES_NDESTROY_RESOLV OR _MONGOC_HAVE_RES_NCLOSE_RESOLV))
+        set(RESOLVE_LIB_NAME resolv)
+    else()
+        # Can we use name resolution with just libc?
+        unset(CMAKE_REQUIRED_LIBRARIES)
+        check_symbol_exists(res_nsearch "${resolve_headers}" _MONGOC_HAVE_RES_NSEARCH_NOLINK)
+        check_symbol_exists(res_search "${resolve_headers}" _MONGOC_HAVE_RES_SEARCH_NOLINK)
+        check_symbol_exists(res_ndestroy "${resolve_headers}" _MONGOC_HAVE_RES_NDESTROY_NOLINK)
+        check_symbol_exists(res_nclose "${resolve_headers}" _MONGOC_HAVE_RES_NCLOSE_NOLINK)
+        if((_MONGOC_HAVE_RES_NSEARCH_NOLINK OR _MONGOC_HAVE_RES_SEARCH_NOLINK)
+            AND (_MONGOC_HAVE_RES_NDESTROY_NOLINK OR _MONGOC_HAVE_RES_NCLOSE_NOLINK))
+            set(resolve_is_libc TRUE)
+            message(VERBOSE "Name resolution is provided by the C runtime")
+        endif()
+    endif()
+endif()
+
+_mongo_pick(MONGOC_HAVE_DNSAPI 1 0 _MONGOC_HAVE_DNSAPI)
+_mongo_pick(MONGOC_HAVE_RES_NSEARCH 1 0 [[_MONGOC_HAVE_RES_NSEARCH_NOLINK OR _MONGOC_HAVE_RES_NSEARCH_RESOLV]])
+_mongo_pick(MONGOC_HAVE_RES_SEARCH 1 0 [[_MONGOC_HAVE_RES_SEARCH_NOLINK OR _MONGOC_HAVE_RES_SEARCH_RESOLV]])
+_mongo_pick(MONGOC_HAVE_RES_NDESTROY 1 0 [[_MONGOC_HAVE_RES_NDESTROY_NOLINK OR _MONGOC_HAVE_RES_NDESTROY_RESOLV]])
+_mongo_pick(MONGOC_HAVE_RES_NCLOSE 1 0 [[_MONGOC_HAVE_RES_NCLOSE_NOLINK OR _MONGOC_HAVE_RES_NCLOSE_RESOLV]])
+
+if(RESOLVE_LIB_NAME OR resolve_is_libc)
+    # Define the resolver interface:
+    add_library(_mongoc-resolve INTERFACE)
+    add_library(mongo::detail::c_resolve ALIAS _mongoc-resolve)
```

Comment:
`mongo` is the prefix only because that is the name given when the targets are installed via `install(TARGETS ... NAMESPACE mongo::)`, so it helps that the names match before inside and outside of the source tree. I would prefer it be `mongoc::`, but alas...

Reply:
The `mongo::` namespace was chosen deliberately so that to a consumer of multiple projects, they would all exist in a somewhat sensible way under the same target namespace, like `mongo::mongoc_shared`, `mongo::bsoncxx_static`, and `mongo::mongocrypt`. If these targets (i.e., the resolve target) aren't mean to be user-visible or directly used by consumers, then a different scheme is probably not an issue.